src/main/scala/io/iohk/ethereum/mining/Miner.scala

```diff
@@ -0,0 +1,194 @@
+import scala.concurrent.duration._
+import scala.util.{Failure, Random, Success, Try}
+
+class Miner(
+    appStateStorage: AppStateStorage,
+    blockGenerator: BlockGenerator,
+    ommersPool: ActorRef,
+    pendingTransactionsManager: ActorRef,
+    syncController: ActorRef,
+    miningConfig: MiningConfig)
+  extends Actor with ActorLogging {
+
+  import Miner._
+  import akka.pattern.ask
+
+  var currentEpoch: Option[Long] = None
+  var currentEpochDagSize: Option[Long] = None
+  var currentEpochDag: Option[Array[Array[Int]]] = None
+
+  override def receive: Receive = stopped
+
+  def stopped: Receive = {
+    case StartMining =>
+      context become started
+      self ! ProcessMining
+    case ProcessMining => // nothing
+  }
+
+  def started: Receive = {
+    case StopMining => context become stopped
+    case ProcessMining => processMining()
+  }
+
+  def processMining(): Unit = {
+    val blockNumber = appStateStorage.getBestBlockNumber() + 1
+    val epoch = Ethash.epoch(blockNumber.toLong)
+
+    val (dag, dagSize) = (currentEpoch, currentEpochDag, currentEpochDagSize) match {
+      case (Some(`epoch`), Some(dag), Some(dagSize)) => (dag, dagSize)
+      case _ =>
+        val seed = Ethash.seed(epoch)
+        val dagSize = Ethash.dagSize(epoch)
+        val dag =
+          if (!dagFile(seed).exists()) generateDagAndSaveToFile(epoch, dagSize, seed)
+          else loadDagFromFile(seed, dagSize)
+
+        currentEpoch = Some(epoch)
+        currentEpochDag = Some(dag)
+        currentEpochDagSize = Some(dagSize)
+      (dag, dagSize)
+    }
+
+    getBlockForMining(blockNumber) onComplete {
+      case Success(PendingBlock(block, _)) =>
+        val headerHash = crypto.kec256(BlockHeader.getEncodedWithoutNonce(block.header))
+        val mineResult = mine(headerHash, block.header.difficulty.toLong, dagSize, dag, miningConfig.mineRounds)
+        mineResult.foreach { case (pow, nonce) =>
+          syncController ! RegularSync.MinedBlock(block.copy(header = block.header.copy(nonce = nonce, mixHash = pow.mixHash)))
+        }
+        self ! ProcessMining
+
+      case Failure(ex) =>
+        log.error("Unable to get block for mining", ex)
+        context.system.scheduler.scheduleOnce(10.seconds, self, ProcessMining)
+    }
+  }
+
+  private def dagFile(seed: ByteString): File = {
+    new File(s"${miningConfig.ethashDir}/full-R${Ethash.Revision}-${Hex.toHexString(seed.take(8).toArray[Byte])}")
+  }
+
+  private def generateDagAndSaveToFile(epoch: Long, dagSize: Long, seed: ByteString): Array[Array[Int]] = {
+    // scalastyle:off magic.number
+
+    val file = dagFile(seed)
+    if (file.exists()) file.delete()
+    file.createNewFile()
+
+    val outputStream = new FileOutputStream(dagFile(seed).getAbsolutePath)
+    outputStream.write(Array(0xfe, 0xca, 0xdd, 0xba, 0xad, 0xde, 0xe1, 0xfe).map(_.toByte))
+
+    val cache = Ethash.makeCache(epoch)
+    val res = new Array[Array[Int]]((dagSize / Ethash.HASH_BYTES).toInt)
```

Comment:
This expr `dagSize / Ethash.HASH_BYTES).toInt` seems to be used in 3 places, maybe we can precompute it earlier and pass it to function instead of `dagSize` ?